rofl-scheduler/src/manager.rs

```diff
@@ -0,0 +1,1331 @@
+            .query(
+                round,
+                "roflmarket.Instances",
+                market::types::ProviderQuery {
+                    provider: self.cfg.provider_address,
+                },
+            )
+            .await?;
+
+        let mut accepted: BTreeMap<InstanceId, Instance> = BTreeMap::new();
+        let mut pending_start: Vec<(Instance, Deployment, bool)> = Vec::new();
+        let mut pending_stop: Vec<(InstanceId, bool)> = Vec::new();
+        let mut instance_updates: BTreeMap<InstanceId, InstanceUpdates> = BTreeMap::new();
+        let mut maybe_remove: Vec<(InstanceId, u64)> = Vec::new();
+        let mut claim_payment: Vec<InstanceId> = Vec::new();
+        let mut resources_used: Resources = Default::default();
+
+        for instance in instances {
+            // Remove known instances.
+            running_unknown.remove(&instance.id);
+
+            match instance.status {
+                InstanceStatus::Created => {
+                    // Instance has not yet been accepted, nothing to do.
+                    continue;
+                }
+                InstanceStatus::Cancelled => {
+                    // Instance has been cancelled, make sure it is stopped if running.
+                    if running.contains_key(&instance.id) {
+                        pending_stop.push((instance.id, true));
+                    }
+                    maybe_remove.push((instance.id, instance.updated_at));
+                    continue;
+                }
+                InstanceStatus::Accepted => {
+                    // Instance has been accepted, check if we should be hosting it.
+                    if instance.node_id.unwrap() != local_node_id {
+                        continue;
+                    }
+                }
+            }
+
+            accepted.insert(instance.id, instance.clone());
+
+            // Check if the instance is still paid for. If not, we immediately stop it and schedule
+            // its removal.
+            if instance.paid_until < now {
+                slog::info!(self.logger, "instance not paid for, stopping";
+                    "id" => ?instance.id,
+                );
+
+                if running.contains_key(&instance.id) {
+                    pending_stop.push((instance.id, true));
+                }
+                maybe_remove.push((instance.id, instance.paid_until));
+                continue;
+            }
+
+            // Compute total provisioned resources.
+            resources_used.add(&instance.resources);
+
+            // Discover any pending commands to see if there is a "deploy" command somewhere in
+            // there. This allows us to immediately deploy the right thing instead of first
+            // deploying an old version and then immediately upgrading.
+            let cmds: Vec<market::types::QueuedCommand> = client
+                .query(
+                    round,
+                    "roflmarket.InstanceCommands",
```

Comment:
Market calls could be moved to a special struct (e.g. client), which would be initialized on every (new) round and called like `market.InstanceCommands(provider, instance.id).await?`.

Reply:
Good idea.